### PR TITLE
Wrap errors, encode signature Subject/Issuer.

### PR DESCRIPTION
**Pull request by mattmoor**
As I was poking around the code base preparing for the upstream wrapped error change to land, I refactored a bunch of code paths (in part preparing to integrate that change).

There are two main changes:
1. Pull in the wrapped error changes,
2. Drop the use of `PLACEHOLDER` and encode the actual `Subject`/`Issuer` from the `oci.Signature`.

There are a bunch of minor / supporting changes:
 - collapse several redundant code paths,
 - pre-allocate several arrays and maps to avoid resizing,
 - make some of the error messages more consistent across sig/att and key/keyless,
 - use the wrapped errors a handful of places.

Fixes: https://github.com/sigstore/policy-controller/issues/109

### CODE REVIEW

**Comment by mattmoor** [comment]
cc @vaikas

**Comment by mattmoor** [comment]
FML, hopefully this lands: https://github.com/google/certificate-transparency-go/pull/962

**Comment by mattmoor** [comment]
I've added a `replace` for @jdolitsky change here to see if CI finds any other problems, but we need that change to land, or to do some work in `sigstore/cosign` to unblock the flow of changes 😞

File: pkg/webhook/validation.go
Comment (mattmoor) [inline pkg/webhook/validation.go:45]: It turns out that when `sps` is empty, this always returns an error, so the flow we had was pointless.

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:473]: This is extracted directly from how `cosign` outputs this information during verification today.

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:503]: As I said above, this is unreachable today.

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:520]: same

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:570]: Similar to the API, I thought it prudent to avoid the term "fulcio" in error messages, especially since the API keyword is `keyless`.  I also tried to align this with the more detailed errors for the `key` path above.

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:589]: Pretty sure this too is unreachable, but since we have the loop and all I opted for wrapping the same error here instead.

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:389]: This was necessary because of the `default` case below, which errors when both `result.signatures` and `result.attestations` were empty.  I've changed that to allow this case, so that we don't have to look for these non-`const` sentinel strings.

File: pkg/webhook/validator.go
Comment (mattmoor) [inline pkg/webhook/validator.go:436]: This is the `default` case I was referring to above.  We hit this when there wasn't an error, but there also weren't any validates `signatures` or `attestations`, which corresponds to the `static: { action: pass }` path.

**Review by mattmoor — COMMENTED** [review]
Left some comments inline to explain things, which probably wouldn't make sense as committed code comments.

File: pkg/webhook/validator.go
Comment (vaikas) [inline pkg/webhook/validator.go:534]: while changing this, mind fixing the comment above that I failed to update :)

File: pkg/webhook/validator.go
Comment (vaikas) [inline pkg/webhook/validator.go:564]: looks like maybe copy&paste from signature validation, can you add something here about the attestation.

File: pkg/webhook/validator.go
Comment (vaikas) [inline pkg/webhook/validator.go:579]: ditto.

File: pkg/webhook/validator.go
Comment (vaikas) [inline pkg/webhook/validator.go:419]: just for my own edification, is that really the only reason why that would return !ok?

File: pkg/webhook/validator.go
Comment (vaikas) [reply pkg/webhook/validator.go:389]: I see what you're doing here but having a default below that is fail open just makes me a bit nervous (not in this PR), but in general.
What do you think about changing this structure here and add a field for Static pass instead:
https://github.com/sigstore/policy-controller/blob/main/pkg/webhook/validator_result.go#L40

Say, AuthorityMatch.StaticPass

Then we could tag this as having satisfied that and check for that before the default clause and leave default be fail?

Just thinking again for future and safety?

File: pkg/webhook/validation.go
Comment (vaikas) [reply pkg/webhook/validation.go:45]: superduper.

File: pkg/webhook/validator.go
Comment (vaikas) [inline pkg/webhook/validator.go:33]: We've been trying (with the sigstore refactoring work that @imjasonh and others have been doing) to reduce dependencies to cosign and where ever possible to use sigstore/sigstore, so I'm not sure we want to add this one here. I can also see us maybe having additional error types that we define so maybe we could just have our Errors that we return just like you did in sigstore/cosign and use them here instead.

**Review by vaikas — COMMENTED** [review]
Thanks for doing this, great find on the sps never being 0 👍 Also, thanks for filling in the placeholder!

File: pkg/webhook/validator.go
Comment (mattmoor) [reply pkg/webhook/validator.go:33]: I get that, but here it is a little chicken/egg.  I only added the error stuff to cosign because that's where the logic still lives (and that should be trivial to move to `sigstore/sigstore` when the signing/verification moves).

I'd have to look at what it would take to move the extension stuff over, but it'd be great to have a library in `sigstore/sigstore` that all of `cosign`, `fulcio`, and `policy-controller` can lean on 🤞

File: pkg/webhook/validator.go
Comment (mattmoor) [reply pkg/webhook/validator.go:389]: As long as we're not lying about it having signatures, I'm happy 😁

File: pkg/webhook/validator.go
Comment (mattmoor) [reply pkg/webhook/validator.go:421]: Yup:
image

File: pkg/webhook/validator.go
Comment (mattmoor) [reply pkg/webhook/validator.go:539]: Yup, done!

File: pkg/webhook/validator.go
Comment (mattmoor) [reply pkg/webhook/validator.go]: I prefixed the signature paths with `signature` and the attestation paths with `attestation` to distinguish.  I guess I made them too similar 🤣

File: pkg/webhook/validator.go
Comment (vaikas) [reply pkg/webhook/validator.go:389]: Yeah, that's why think having a new field that's not lying is good. I probably should've just added it the first time around :)

File: pkg/webhook/validator.go
Comment (vaikas) [reply pkg/webhook/validator.go:421]: Thanks lazygithub :)

File: pkg/webhook/validator.go
Comment (vaikas) [reply pkg/webhook/validator.go:33]: Ack, that makes sense so that when the signing moves the errors would move as well. Hindsight, maybe the error stuff should've been put into sigstore/sigstore to begin with?

Comment (mattmoor) [reply pkg/webhook/validator.go:33]: Possibly, but I opted for the faster/lazier version of treating it as part of the verification chunk that'll move later 😏

File: pkg/webhook/validation.go
Comment (hectorj2f) [reply pkg/webhook/validation.go:45]: 👍🏻

File: pkg/webhook/validator.go
Comment (hectorj2f) [reply pkg/webhook/validator.go:503]: yes, it makes sense.

**Comment by vaikas** [comment]
LGTM, thanks much!

**Comment by hectorj2f** [comment]
lgtm, thanks for the PR.

**Comment by mattmoor** [comment]
NP, I'll rebase once Josh's PR merges 🤞

**Comment by mattmoor** [comment]
Just rebased on @jdolitsky change 🙏

**Review by jdolitsky — APPROVED** [review]

**Review by vaikas — APPROVED** [review]

**Pull request merged**
